File: build_result/Intel/VAM-RPC.app/Contents/Resources/agent.ts
// agent.ts
import { Client, type Activity } from "https://deno.land/x/discord_rpc@0.3.2/mod.ts";

const SUPPORT_DIR = `${Deno.env.get("HOME")}/Library/Application Support/VAM-RPC`;
const STATUS_PATH = `${SUPPORT_DIR}/status.txt`;
const CONFIG_PATH = `${SUPPORT_DIR}/data/config.json`;
//heyyy, here! you can change the icon here if you want different image! maybe in the future we can add custom image support?
const ICONS = {
    PLAY: "https://i.imgur.com/6uuaC8A.png",
    PAUSE: "https://i.imgur.com/8oAUykh.png",
    APPLE_MUSIC: "https://upload.wikimedia.org/wikipedia/commons/thumb/5/5f/Apple_Music_icon.svg/512px-Apple_Music_icon.svg.png"
};

interface Settings {
    refreshInterval: number;
    activityName: string;
    enableSpotifyButton: boolean;
    spotifyButtonLabel: string;
    enableAppleMusicButton: boolean;
    appleMusicButtonLabel: string;
    enableSonglinkButton: boolean;
    songlinkButtonLabel: string;
    enableYoutubeMusicButton: boolean;
    youtubeMusicButtonLabel: string;
    detailsString: string;
    stateString: string;
    largeImageText: string;
    smallImageText: string;
    smallImageSource: string;
}

async function loadSettings(): Promise<Settings> {
    const defaultSettings: Settings = {
        refreshInterval: 5,
        activityName: "Apple Music",
        enableSpotifyButton: true,
        spotifyButtonLabel: "♫ Find On Spotify",
        enableAppleMusicButton: true,
        appleMusicButtonLabel: "♫ Open on  Music",
        enableSonglinkButton: false, 
        songlinkButtonLabel: "♫ Find on Songlink",
        enableYoutubeMusicButton: false, 
        youtubeMusicButtonLabel: "♫ Find on YT Music",
        detailsString: "{name}",
        stateString: "by {artist}",
        largeImageText: "{name} - {album}",
        smallImageText: "{artist}",
        smallImageSource: "default",
    };

    try {
        const content = await Deno.readTextFile(CONFIG_PATH);
        const loaded = { ...defaultSettings, ...JSON.parse(content) };
        if (!["default", "albumArt", "artistArt", "playbackStatus", "appIcon"].includes(loaded.smallImageSource)) {
            loaded.smallImageSource = "default";
        }
        return loaded;
    } catch (e) {
        console.warn(`VAM-RPC Agent: ⚠️ Could not load settings. Using defaults.`);
        return defaultSettings;
    }
}

async function runJsInMusic<T>(script: string): Promise<T> {
    const wrappedScript = `JSON.stringify( (() => { ${script} })() )`;
    const cmd = new Deno.Command("osascript", { args: ["-l", "JavaScript", "-e", wrappedScript] });
    const { stdout, code, stderr } = await cmd.output();
    if (code !== 0) throw new Error(`AppleScript Error: ${new TextDecoder().decode(stderr)}`);
    return JSON.parse(new TextDecoder().decode(stdout)) as T;
}

interface TrackProps { name: string; artist: string; album: string; duration: number; playerPosition: number; }

async function getMusicState(): Promise<{ state: "playing" | "paused" | "stopped", track?: TrackProps }> {
    const script = `
        const music = Application("Music");
        if (!music.running()) return { state: "stopped" };
        const state = music.playerState();
        
        // Return track data even if paused so we can show "Paused: Song Name"
        if (state === "playing" || state === "paused") {
            try {
                const track = music.currentTrack;
                return {
                    state: state,
                    track: {
                        name: track.name(), artist: track.artist(), album: track.album(),
                        duration: track.duration(), playerPosition: music.playerPosition()
                    }
                };
            } catch (e) { return { state: "stopped" }; }
        }
        return { state: "stopped" };
    `;
    return await runJsInMusic(script);
}

interface TrackExtras { 
    albumArtUrl?: string; 
    artistArtUrl?: string;
    trackViewUrl?: string; 
}

async function fetchTrackExtras(props: TrackProps): Promise<TrackExtras> {
    const cleanArtist = props.artist.replace(/\(.*\)/g, "").trim();
    const cleanAlbum = props.album.replace(/\(.*\)/g, "").trim();
    
    let albumData = await fetchItunesAlbum(cleanArtist, cleanAlbum);
    
    if (!albumData || !albumData.artwork) {
        const fallbackArt = await fetchDeezerAlbum(cleanArtist, cleanAlbum);
        if (fallbackArt) {
            albumData = { artwork: fallbackArt, url: albumData?.url ?? "" };
        }
    }

    const artistData = await fetchDeezerArtist(cleanArtist);

    return {
        albumArtUrl: albumData?.artwork,
        trackViewUrl: albumData?.url,
        artistArtUrl: artistData
    };
}

async function fetchItunesAlbum(artist: string, album: string): Promise<{ artwork: string, url: string } | null> {
    const query = `${artist} ${album}`;
    const url = `https://itunes.apple.com/search?term=${encodeURIComponent(query)}&entity=album&limit=1`;
    try {
        const resp = await fetch(url);
        if (!resp.ok) return null;
        const json = await resp.json();
        if (json.resultCount > 0) {
            const result = json.results[0];
            return { 
                artwork: result.artworkUrl100.replace("100x100bb", "1000x1000bb"), 
                url: result.collectionViewUrl 
            };
        }
    } catch (e) {}
    return null;
}

async function fetchDeezerArtist(artist: string): Promise<string | undefined> {
    const url = `https://api.deezer.com/search/artist?q=${encodeURIComponent(artist)}&limit=1`;
    try {
        const resp = await fetch(url);
        if (!resp.ok) return undefined;
        const json = await resp.json();
        if (json.data && json.data.length > 0) return json.data[0].picture_xl ?? json.data[0].picture_medium;
    } catch (e) {}
    return undefined;
}

async function fetchDeezerAlbum(artist: string, album: string): Promise<string | undefined> {
    const query = `${artist} ${album}`;
    const url = `https://api.deezer.com/search/album?q=${encodeURIComponent(query)}&limit=1`;
    try {
        const resp = await fetch(url);
        if (!resp.ok) return undefined;
        const json = await resp.json();
        if (json.data && json.data.length > 0) return json.data[0].cover_xl ?? json.data[0].cover_big;
    } catch (e) {}
    return undefined;
}

function ensureValidString(value: string | null | undefined, minLength = 2, maxLength = 128): string {
    if (!value) return " ".repeat(minLength);
    if (value.length < minLength) return value.padEnd(minLength, " ");
    if (value.length > maxLength) return `${value.slice(0, maxLength - 1)}…`;
    return value;
}

function formatString(template: string, track: TrackProps): string {
    return template
        .replace(/\{name\}/g, track.name)
        .replace(/\{artist\}/g, track.artist)
        .replace(/\{album\}/g, track.album);
}

async function main() {
    console.log("VAM-RPC Agent: Starting service...");
    await Deno.writeTextFile(STATUS_PATH, "Service Starting...");
    
    const settings = await loadSettings();
    const rpc = new Client({ id: "773825528921849856" });
    
    try {
        await rpc.connect();
        console.log("VAM-RPC Agent: ✅ RPC Connected.");
    } catch (e) {
        console.error("Failed to connect to Discord RPC:", e);
        await Deno.writeTextFile(STATUS_PATH, "Error: Discord RPC Failed");
        return;
    }

    const tick = async () => {
        try {
            const music = await getMusicState();

            // Handle BOTH playing and paused states
            if ((music.state === "playing" || music.state === "paused") && music.track) {
                const track = music.track;
                const isPaused = music.state === "paused";
                
                await Deno.writeTextFile(STATUS_PATH, isPaused ? `Paused: ${track.name}` : `Playing: ${track.name}`);

                const extras = await fetchTrackExtras(track);
                
                // Only calculate timestamps if playing (Discord automatically hides time left if paused, usually)
                let timestamps;
                if (!isPaused && track.duration > 0) {
                    const now = Date.now();
                    const start = Math.round(now - (track.playerPosition * 1000));
                    const end = Math.round(start + (track.duration * 1000));
                    timestamps = { start, end };
                }

                // DETERMINE SMALL IMAGE SOURCE
                let smallImageKey: string | undefined = undefined;
                let smallImageTextValue = ensureValidString(formatString(settings.smallImageText, track));

                if (settings.smallImageSource === "albumArt" && extras.albumArtUrl) {
                    smallImageKey = extras.albumArtUrl;
                } else if (settings.smallImageSource === "artistArt" && extras.artistArtUrl) {
                    smallImageKey = extras.artistArtUrl;
                } else if (settings.smallImageSource === "playbackStatus") {
                    // Use web URLs instead of asset keys
                    smallImageKey = isPaused ? ICONS.PAUSE : ICONS.PLAY;
                    smallImageTextValue = isPaused ? "Paused" : "Playing";
                } else if (settings.smallImageSource === "appIcon") {
                    // Use web URL
                    smallImageKey = ICONS.APPLE_MUSIC;
                    smallImageTextValue = "Apple Music";
                }

                // DETERMINE LARGE IMAGE
                // If album art is missing, fall back to the generic Apple Music URL (not "appicon")
                const largeImageKey = extras.albumArtUrl ?? ICONS.APPLE_MUSIC;

                const activity: Activity = {
                    type: 2, // Listening
                    name: ensureValidString(formatString(settings.activityName, track)),
                    details: ensureValidString(formatString(settings.detailsString, track)),
                    state: ensureValidString(formatString(settings.stateString, track)),
                    assets: {
                        large_image: largeImageKey,
                        large_text: ensureValidString(formatString(settings.largeImageText, track)),
                        // Only add small image if key exists
                        ...(smallImageKey ? { 
                            small_image: smallImageKey, 
                            small_text: smallImageTextValue 
                        } : {})
                    }
                };
                
                if (timestamps) { activity.timestamps = timestamps; }

                // Buttons
                const buttons = [];
                const searchQuery = encodeURIComponent(`${track.name} ${track.artist}`);

                if (settings.enableAppleMusicButton) {
                    const url = extras.trackViewUrl ? extras.trackViewUrl : `https://music.apple.com/us/search?term=${searchQuery}`;
                    buttons.push({ label: ensureValidString(settings.appleMusicButtonLabel, 2, 32), url });
                }
                if (settings.enableSpotifyButton) {
                    buttons.push({ label: ensureValidString(settings.spotifyButtonLabel, 2, 32), url: `https://open.spotify.com/search/${searchQuery}` });
                }
                if (settings.enableSonglinkButton) {
                    buttons.push({ label: ensureValidString(settings.songlinkButtonLabel, 2, 32), url: `https://song.link/s?q=${searchQuery}` });
                }
                if (settings.enableYoutubeMusicButton) {
                    buttons.push({ label: ensureValidString(settings.youtubeMusicButtonLabel, 2, 32), url: `https://music.youtube.com/search?q=${searchQuery}` });
                }
                if (buttons.length > 0) { activity.buttons = buttons; }

                await rpc.setActivity(activity);

            } else {
                // Stopped
                await rpc.clearActivity();
                await Deno.writeTextFile(STATUS_PATH, "Stopped");
            }
        } catch (e) {
            console.error(`VAM-RPC Agent: Error - ${(e as Error).message}`);
            await Deno.writeTextFile(STATUS_PATH, `Error: ${(e as Error).message}`);
        } finally {
            setTimeout(tick, settings.refreshInterval * 1000);
        }
    };
    
    tick();
}

main().catch(e => {
    console.error(`VAM-RPC Agent: Fatal Error - ${e.message}`);
    Deno.writeTextFile(STATUS_PATH, `Fatal Error: ${e.message}`);
    Deno.exit(1);
});

File: swift/Agent.swift
import Foundation



struct TrackProps {
    var name: String
    var artist: String
    var album: String
    var duration: Double
    var playerPosition: Double
}

struct TrackExtras {
    var albumArtUrl: String?
    var artistArtUrl: String?
    var trackViewUrl: String?
}

class Agent {
    static let shared = Agent()
    private let rpc = DiscordRPC(clientId: "773825528921849856")
    private var timer: Timer?
    private let spinnerApiBase = "https://able-pig-53.deno.dev"
    private var isPaused = false
    
    private var lastTrackKey: String = ""
    private var lastExtras: TrackExtras?
    
    private let icons = [
        "PLAY": "https://i.imgur.com/6uuaC8A.png",
        "PAUSE": "https://i.imgur.com/8oAUykh.png",
        "APPLE_MUSIC": "https://upload.wikimedia.org/wikipedia/commons/thumb/5/5f/Apple_Music_icon.svg/512px-Apple_Music_icon.svg.png"
    ]
    
    func start() {
        isPaused = false
        // Run immediately then schedule
        DispatchQueue.global(qos: .background).async {
            self.tick()
        }
    }
    
    func stop() {
        timer?.invalidate()
        timer = nil
        rpc.clearActivity()
        rpc.disconnect()
        writeStatus("Stopped")
    }
    
    func togglePause() -> Bool {
        isPaused.toggle()
        if isPaused {
            timer?.invalidate()
            timer = nil
            rpc.clearActivity()
            writeStatus("RPC Paused")
            return true // is Paused
        } else {
            start()
            return false // is not Paused
        }
    }
    
    private func loadSettings() -> Settings {
        let path = (NSString(string: "~/Library/Application Support/VAM-RPC/data/config.json").expandingTildeInPath)
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
              let settings = try? JSONDecoder().decode(Settings.self, from: data) else {
            return Settings.defaultSettings()
        }
        return settings
    }
    
    private func writeStatus(_ status: String) {
        let path = (NSString(string: "~/Library/Application Support/VAM-RPC/status.txt").expandingTildeInPath)
        try? status.write(toFile: path, atomically: true, encoding: .utf8)
    }
    
    private func runAppleScript(_ script: String) -> String? {
        return autoreleasepool {
            var error: NSDictionary?
            if let appleScript = NSAppleScript(source: script) {
                let result = appleScript.executeAndReturnError(&error)
                if error == nil {
                    return result.stringValue
                } else {
                    writeStatus("AppleScript Error")
                    print("AppleScript Error: \(String(describing: error))")
                }
            }
            return nil
        }
    }
    
    private func getMusicState() -> (state: String, track: TrackProps?) {
        let script = """
        tell application "System Events"
            if not (exists process "Music") then return "stopped"
        end tell
        tell application "Music"
            if player state is stopped then return "stopped"
            set myState to player state as string
            set tName to name of current track
            set tArtist to artist of current track
            set tAlbum to album of current track
            set tDur to duration of current track
            set tPos to player position
            return myState & "|||" & tName & "|||" & tArtist & "|||" & tAlbum & "|||" & tDur & "|||" & tPos
        end tell
        """
        
        guard let result = runAppleScript(script) else { return ("stopped", nil) }
        if result == "stopped" { return ("stopped", nil) }
        
        let parts = result.components(separatedBy: "|||")
        if parts.count == 6 {
            let state = parts[0]
            let durationStr = parts[4].replacingOccurrences(of: ",", with: ".")
            let posStr = parts[5].replacingOccurrences(of: ",", with: ".")
            let track = TrackProps(
                name: parts[1],
                artist: parts[2],
                album: parts[3],
                duration: Double(durationStr) ?? 0,
                playerPosition: Double(posStr) ?? 0
            )
            return (state, track)
        }
        return ("stopped", nil)
    }
    
    private func tick() {
        if isPaused { return }
        
        let settings = loadSettings()
        let music = getMusicState()
        
        if (music.state == "playing" || music.state == "paused"), let track = music.track {
            let trackIsPaused = music.state == "paused"
            writeStatus(trackIsPaused ? "Paused: \(track.name)" : "Playing: \(track.name)")
            
            let trackKey = "\(track.name) - \(track.artist) - \(track.album)"
            if trackKey == self.lastTrackKey, let extras = self.lastExtras {
                self.updateDiscord(track: track, settings: settings, extras: extras, isPaused: trackIsPaused)
            } else {
                fetchTrackExtras(track: track) { extras in
                    self.lastTrackKey = trackKey
                    self.lastExtras = extras
                    self.updateDiscord(track: track, settings: settings, extras: extras, isPaused: trackIsPaused)
                }
            }
        } else {
            rpc.clearActivity()
            writeStatus("Stopped")
            scheduleNextTick(interval: settings.refreshInterval)
        }
    }
    
    private func scheduleNextTick(interval: Int) {
        DispatchQueue.main.async {
            self.timer?.invalidate()
            self.timer = Timer.scheduledTimer(withTimeInterval: TimeInterval(interval), repeats: false) { _ in
                DispatchQueue.global(qos: .background).async { self.tick() }
            }
        }
    }
    
    private func updateDiscord(track: TrackProps, settings: Settings, extras: TrackExtras, isPaused: Bool) {
        var activity: [String: Any] = [
            "type": 2,
            "name": self.formatString(settings.activityName, track),
            "details": self.formatString(settings.detailsString, track),
            "state": self.formatString(settings.stateString, track)
        ]
        
        if !isPaused && track.duration > 0 {
            let now = Int(Date().timeIntervalSince1970)
            let start = now - Int(track.playerPosition)
            let end = start + Int(track.duration)
            activity["timestamps"] = ["start": start, "end": end]
        }
        
        var smallImageKey: String?
        var smallImageTextValue = self.formatString(settings.smallImageText, track)
        
        if settings.smallImageSource == "albumArt", let url = extras.albumArtUrl {
            smallImageKey = url
        } else if settings.smallImageSource == "artistArt", let url = extras.artistArtUrl {
            smallImageKey = url
        } else if settings.smallImageSource == "playbackStatus" {
            smallImageKey = isPaused ? self.icons["PAUSE"] : self.icons["PLAY"]
            smallImageTextValue = isPaused ? "Paused" : "Playing"
        } else if settings.smallImageSource == "appIcon" {
            smallImageKey = self.icons["APPLE_MUSIC"]
            smallImageTextValue = "Apple Music"
        } else if settings.smallImageSource == "artistArtAnimated", let url = extras.artistArtUrl {
            if let encoded = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
                smallImageKey = "\(self.spinnerApiBase)/spin.gif?url=\(encoded)"
            }
        } else if settings.smallImageSource == "albumArtAnimated", let url = extras.albumArtUrl {
            if let encoded = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
                smallImageKey = "\(self.spinnerApiBase)/spin.gif?url=\(encoded)"
            }
        }
        
        var assets: [String: String] = [
            "large_image": extras.albumArtUrl ?? self.icons["APPLE_MUSIC"]!,
            "large_text": self.formatString(settings.largeImageText, track)
        ]
        if let sm = smallImageKey {
            assets["small_image"] = sm
            assets["small_text"] = smallImageTextValue
        }
        activity["assets"] = assets
        
        var buttons = [[String: String]]()
        let searchQuery = "\(track.name) \(track.artist)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        
        if settings.enableAppleMusicButton {
            let url = extras.trackViewUrl ?? "https://music.apple.com/us/search?term=\(searchQuery)"
            buttons.append(["label": self.formatString(settings.appleMusicButtonLabel, track), "url": url])
        }
        if settings.enableSpotifyButton {
            buttons.append(["label": self.formatString(settings.spotifyButtonLabel, track), "url": "https://open.spotify.com/search/\(searchQuery)"])
        }
        if settings.enableSonglinkButton {
            buttons.append(["label": self.formatString(settings.songlinkButtonLabel, track), "url": "https://song.link/s?q=\(searchQuery)"])
        }
        if settings.enableYoutubeMusicButton {
            buttons.append(["label": self.formatString(settings.youtubeMusicButtonLabel, track), "url": "https://music.youtube.com/search?q=\(searchQuery)"])
        }
        
        if buttons.count > 2 { buttons = Array(buttons.prefix(2)) }
        if !buttons.isEmpty { activity["buttons"] = buttons }
        
        self.rpc.setActivity(activity: activity)
        
        self.scheduleNextTick(interval: settings.refreshInterval)
    }
    
    private func formatString(_ template: String, _ track: TrackProps) -> String {
        var str = template
            .replacingOccurrences(of: "{name}", with: track.name)
            .replacingOccurrences(of: "{artist}", with: track.artist)
            .replacingOccurrences(of: "{album}", with: track.album)
        if str.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return "  " }
        if str.count < 2 { str = str.padding(toLength: 2, withPad: " ", startingAt: 0) }
        if str.count > 128 { str = String(str.prefix(127)) + "…" }
        return str
    }
    
    private func fetchTrackExtras(track: TrackProps, completion: @escaping (TrackExtras) -> Void) {
        let fullArtist = track.artist.replacingOccurrences(of: "\\(.*\\)", with: "", options: .regularExpression).trimmingCharacters(in: .whitespaces)
        let cleanAlbum = track.album.replacingOccurrences(of: "\\(.*\\)", with: "", options: .regularExpression).trimmingCharacters(in: .whitespaces)
        
        let primaryArtist = fullArtist.components(separatedBy: CharacterSet(charactersIn: ",&")).first?
            .components(separatedBy: " feat. ").first?
            .components(separatedBy: " ft. ").first?
            .components(separatedBy: " x ").first?
            .components(separatedBy: " vs ").first?.trimmingCharacters(in: .whitespaces) ?? fullArtist
        
        var extras = TrackExtras()
        let group = DispatchGroup()
        
        group.enter()
        fetchItunesAlbum(artist: fullArtist, album: cleanAlbum) { itunesData in
            if let data = itunesData {
                extras.albumArtUrl = data.artwork
                extras.trackViewUrl = data.url
                group.leave()
            } else {
                self.fetchItunesAlbum(artist: primaryArtist, album: cleanAlbum) { itunesData2 in
                    if let data = itunesData2 {
                        extras.albumArtUrl = data.artwork
                        extras.trackViewUrl = data.url
                    } else {
                        let dGroup = DispatchGroup()
                        dGroup.enter()
                        self.fetchDeezerAlbum(artist: fullArtist, album: cleanAlbum) { cover in
                            if let c = cover {
                                extras.albumArtUrl = c
                                dGroup.leave()
                            } else {
                                self.fetchDeezerAlbum(artist: primaryArtist, album: cleanAlbum) { cover2 in
                                    extras.albumArtUrl = cover2
                                    dGroup.leave()
                                }
                            }
                        }
                        dGroup.wait()
                    }
                    group.leave()
                }
            }
        }
        
        group.enter()
        fetchDeezerArtist(artist: primaryArtist) { picture in
            extras.artistArtUrl = picture
            group.leave()
        }
        
        group.notify(queue: .global(qos: .background)) {
            completion(extras)
        }
    }
    
    private func fetchItunesAlbum(artist: String, album: String, completion: @escaping ((artwork: String, url: String)?) -> Void) {
        let query = "\(artist) \(album)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        let urlStr = "https://itunes.apple.com/search?term=\(query)&entity=album&limit=1"
        guard let url = URL(string: urlStr) else { completion(nil); return }
        
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let results = json["results"] as? [[String: Any]],
                  let first = results.first,
                  let artworkUrl100 = first["artworkUrl100"] as? String,
                  let collectionViewUrl = first["collectionViewUrl"] as? String else {
                completion(nil)
                return
            }
            let hqArtwork = artworkUrl100.replacingOccurrences(of: "100x100bb", with: "1000x1000bb")
            completion((hqArtwork, collectionViewUrl))
        }.resume()
    }
    
    private func fetchDeezerArtist(artist: String, completion: @escaping (String?) -> Void) {
        let query = artist.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        let urlStr = "https://api.deezer.com/search/artist?q=\(query)&limit=1"
        guard let url = URL(string: urlStr) else { completion(nil); return }
        
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let dataArr = json["data"] as? [[String: Any]],
                  let first = dataArr.first else {
                completion(nil)
                return
            }
            let pic = (first["picture_xl"] as? String) ?? (first["picture_medium"] as? String)
            completion(pic)
        }.resume()
    }
    
    private func fetchDeezerAlbum(artist: String, album: String, completion: @escaping (String?) -> Void) {
        let query = "\(artist) \(album)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        let urlStr = "https://api.deezer.com/search/album?q=\(query)&limit=1"
        guard let url = URL(string: urlStr) else { completion(nil); return }
        
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let dataArr = json["data"] as? [[String: Any]],
                  let first = dataArr.first else {
                completion(nil)
                return
            }
            let cover = (first["cover_xl"] as? String) ?? (first["cover_big"] as? String)
            completion(cover)
        }.resume()
    }
}
